qiskit/providers/options.py

```diff
@@ -65,13 +100,47 @@ class Options:
 
     __slots__ = ("_fields", "validator")
 
+    # implementation of the MutableMapping ABC:
+
+    def __getitem__(self, key):
+        return self._fields[key]
+
+    def __setitem__(self, key, value):
+        self.update_options(**{key: value})
+
+    def __delitem__(self, key):
+        del self._fields[key]
```

Comment:
Hmm, I know this is part of the mutable mapping abstract interface so it needs to be implemented (which is why I skipped over this in my earlier review). But on reflection I'm not sure deletion is actually part of the workflow for this class. The typical workflow is the provider package creates the `Options` object at backend initialization and it contains the canonical set of fields for users to set along with their default values. Deletion here doesn't make a ton of sense because nothing should ever be deleting the fields after it's created as they contain the current default behavior for running on a backend.

I'm not 100% sure the best way to proceed with this. My first thought would be to just make this raise an exception. But, that's not great either because it would show methods like `pop()` and `clear()` as being available which they shouldn't be. I'm thinking the best path forward might be to change `MutableMapping` to just `Mapping` and keep the `__setitem__` definition so we support assignment. What do you think?

Reply:
Thanks for the review and apologies for the delayed feedback.

I feel like the actual usage of `Options` is rather that of a (non-frozen) dataclass than a dict, isn't it? Option fields are defined by the backend implementation but the values can be edited. This would however be a quite massive change I guess.

Implementing `__setitem__` however allows to dynamically add options (potentially by mistake in case of a typo), and by symmetry, it would be surprising not to implement the full `MutableMapping` interface but something in-between the standard ABCs. To keep the learning curve smooth, I think that it's easier to say that `Options` is a `MutableMapping` with optional custom validators and no extra peculiarities.

I'd therefore rather be in favor of implementing the full `MutableMapping` interface, in order not to surprise users. Do you find this reasonable?

Reply:
I don't think it's surprising if we say "it fulfills `Mapping`, but you can also change the values" - it can fulfil part of one interface without fulfilling all of the next one in the stack, just like `numpy.array` fulfils `Iterable` but not `Sequence` (because they don't want to implement the named methods that `Sequence` requires).

Reply:
As in, there's nothing about implementing only `Mapping` that means you _can't_ be mutable.  It just means that you provide a particular ABC's definition of "view" semantics around your keys and values.

Reply:
Fair enough. I updated the branch to have `Options` implement `Mapping`.

> it fulfills Mapping, but you can also change the values

and add keys. This is why I find it strange to keep to asymmetry by not providing an interface to delete keys. But I see that it's not the wanted usage of the datastructure.